src/librustc_data_structures/macros.rs

```diff
@@ -0,0 +1,21 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+/// A simple static assertion macro. The first argument should be a unique
+/// ALL_CAPS identifier that describes the condition.
+#[macro_export]
+macro_rules! static_assert {
+    ($name:ident: $test:expr) => {
+        // Use the bool to access an array such that if the bool is false, the access
+        // is out-of-bounds.
+        #[allow(dead_code)]
+        static $name: () = [()][!$test as usize];
```

Comment:
(not really relevant to this PR) So I don't forget: this should have `($test: bool)` to avoid accidentally allowing an integer expression (cc @oli-obk).

Reply:
Just came across this, and I'm curious to know where I can follow the status of having types as macro input specifiers. This would be nice for my [`const_assert!`](https://docs.rs/static_assertions/0.3.1/static_assertions/macro.const_assert.html) macro.

Reply:
I'm not sure what you mean. I am talking about writing `[()][!($test: bool) as usize]`.